### DISCUSSION TITLE
Is there way to use like useReducer ?

### DISCUSSION THREAD

**Discussion by Kavan72** [47ng/nuqs#462] [Q&A] [answered]
At some points, we have many states, so there we are using the `useReducer` approach. Check this:

```ts
interface State {
  productData: Array<Product>;
  isLoading?: boolean;
}

const [state, setState] = useReducer(
  (prev: State, next: Partial<State>): State => {
    return { ...prev, ...next };
  },
  {
    productData: [],
    isLoading: false,
  }
);
```
If you want to update the state, you can do it like this:

```ts
setState({ isLoading: true });
```

And to get a particular state:
```ts
state.isLoading;
```

is it possible with `useQueryState` ?

**Comment by franky47** [comment, marked as answer]
You may do this with `useQueryStates` (plural), which supports multiple related search params.

It does not behave exactly like useReducer, in the sense that it doesn't support a reducer function nor provides a `dispatch` method, but it allows obtaining the previous state and return partial updates:

```ts


const [{ productData, isLoading }, setState] = useQueryStates({
  productData: parseAsJson<Product[]>().withDefault([]),
  isLoading: parseAsBoolean.withDefault(false)
})

setState((previousState) => {
  // Return a partial update of the state type to update the relevant values
  // Example:
  return {
    isLoading: !previousState.isLoading
    // This won't touch the productData
  }
})
```

**Comment by Kavan72** [comment]
Ohh, that's nice. i'll check that and let you know @franky47.

**Reply by Kavan72** [reply]
Btw @franky47 how can i type the full obj, like i'm using this way 
```
const [state, setState] = useQueryStates({
  productData: parseAsJson<Product[]>().withDefault([]),
  isLoading: parseAsBoolean.withDefault(false)
})
```
and i want to type `state` and `setState` so i can pass to another component.

**Reply by franky47** [reply]
What are you trying to obtain as an output type? The type of `state` here should be `{ productData: Product[], isLoading: boolean }`.

**Reply by Kavan72** [reply]
yes, for state is fine like 
```
export interface Props {
  state: { productData: Product[], isLoading: boolean }
  setState: ?
}
```
but what about `setState` ?

**Reply by franky47** [reply]
Ah I see, the best way would be to infer those types directly from the useQueryStates type:

```ts
const searchParamsParsers = {
  productData: parseAsJson<Product[]>().withDefault([]),
  isLoading: parseAsBoolean.withDefault(false)
}
type SearchParamsState = ReturnType<typeof useQueryStates<typeof searchParamsParsers>>[0]
type SearchParamsSetter = ReturnType<typeof useQueryStates<typeof searchParamsParsers>>[1]

type ComponentProps = {
  state: SearchParamsState
  setState: SearchParamsSetter
}

const [state, setState] = useQueryStates(searchParamsParsers)
```

This way the component props stay in sync with whatever key/values you add or change in the `searchParamsParsers` object.

**Reply by Kavan72** [reply]
@franky47 working but this is annoying when you have diff file where you have put all your types

**Comment by Kavan72** [comment]
@franky47 i guess there is some problem when use `parseAsArrayOf`
1) when i update array value, it's render twice 
```
setFilter(previousState => {
  return {
    leadStatus: [...previousState.leadStatus, value]
  }
})
```
2) `clearOnDefault: true` is not working with `parseAsArrayOf`

**Reply by franky47** [reply]
The double render is expected: the first one is when the internal state is updated, and the second when the URL has finished updating (`history.replaceState` is a deferred update, and Next.js hooks like `useSearchParams` will re-render when the URL changes).

For your second point, this is a bug indeed, due to a referential equality check against the default value, which doesn't work for arrays and objects. See PR #504 for a fix.

**Reply by Kavan72** [reply]
Thanks you for quick fix, Can't wait to test with latest release 🙌. 

For double rendering, it's weird because it's only happening with array types.

**Reply by franky47** [reply]
Fix released in 1.17.1.

**Reply by Kavan72** [reply]
@franky47 Is there a way to avoid double rendering? It's causing the API to be called twice for me.

**Reply by franky47** [reply]
Could you open a dedicated issue for this with reproduction steps and your exact setup please?

**Reply by Kavan72** [reply]
created https://github.com/47ng/nuqs/issues/516
